src/components/feedback/Dialog.tsx

```diff
@@ -53,46 +66,42 @@ const DialogNext = function Dialog({
   children,
   initialFocus = 'auto',
   restoreFocus = false,
+  transitionComponent: TransitionComponent,
 
   classes,
   elementRef,
 
   // Forwarded to Panel
   buttons,
   icon,
-  onClose,
+  onClose = noop,
   paddingSize = 'md',
   scrollable = true,
   title,
 
   ...htmlAttributes
 }: DialogProps) {
   const modalRef = useSyncedRef(elementRef);
-  const closeHandler = onClose ?? noop;
   const restoreFocusEl = useRef<HTMLElement | null>(
     document.activeElement as HTMLElement | null
   );
+  const [transitionComponentVisible, setTransitionComponentVisible] =
+    useState(false);
 
-  useClickAway(modalRef, closeHandler, {
-    enabled: closeOnClickAway,
-  });
-
-  useKeyPress(['Escape'], closeHandler, {
-    enabled: closeOnEscape,
-  });
-
-  useFocusAway(modalRef, closeHandler, {
-    enabled: closeOnFocusAway,
-  });
-
-  const dialogDescriptionId = useUniqueId('dialog-description');
+  const closeHandler = useCallback(() => {
+    if (TransitionComponent) {
+      setTransitionComponentVisible(false);
+    } else {
+      onClose();
+    }
+  }, [onClose, TransitionComponent]);
 
-  useEffect(() => {
+  const setInitialFocus = useCallback(() => {
```

Comment:
Although the name of this function is totally accurate, you could consider renaming it something more contextual, like `initializeDialog`. That would represent it as a sort of bag of operations that get executed once when the Dialog is mounted and "ready" (what "ready" means differs based on whether a `TransitionComponent` was provided) and might make the code a little easier to scan. OTOH, this really does just deal in initial focus, so perhaps the abstraction is not helpful...

Reply:
I like the idea. It will quickly help realize that anything that needs to happen "at the beginning", should be there, or invoked from there.

Otherwise, if this component evolves, it's easy to end up putting more stuff in the existing `useEffect`, making it work properly only when there's no `TransitionComponent`.

I'll rename it.